service/history/archivalQueueTaskExecutor.go

```diff
@@ -0,0 +1,164 @@
+		shardContext:   shardContext,
+		workflowCache:  workflowCache,
+		logger:         logger,
+		metricsHandler: metricsHandler,
+	}
+}
+
+func (e *archivalQueueTaskExecutor) Execute(
+	ctx context.Context,
+	executable queues.Executable,
+) (tags []metrics.Tag,
+	isActive bool, err error) {
+	task := executable.GetTask()
+	taskType := queues.GetArchivalTaskTypeTagValue(task)
+	tags = []metrics.Tag{
+		getNamespaceTagByID(e.shardContext.GetNamespaceRegistry(), task.GetNamespaceID()),
+		metrics.TaskTypeTag(taskType),
+		metrics.OperationTag(taskType), // for backward compatibility
+	}
+	switch task := task.(type) {
+	case *tasks.ArchiveExecutionTask:
+		err = e.processArchiveExecutionTask(ctx, task)
+	default:
+		err = fmt.Errorf("task with invalid type sent to archival queue: %+v", task)
+	}
+	return tags, true, err
+}
+
+const defaultRetention = 24 * time.Hour * 30
+
+func (e *archivalQueueTaskExecutor) processArchiveExecutionTask(ctx context.Context,
+	task *tasks.ArchiveExecutionTask) (err error) {
+	weContext, release, err := getWorkflowExecutionContextForTask(ctx, e.workflowCache, task)
+	if err != nil {
+		return err
+	}
+	defer func() { release(err) }()
+
+	mutableState, err := weContext.LoadMutableState(ctx)
+	if err != nil {
+		return err
+	}
+	if mutableState == nil || mutableState.IsWorkflowExecutionRunning() {
+		return nil
+	}
+	branchToken, err := mutableState.GetCurrentBranchToken()
+	if err != nil {
+		return err
+	}
+
+	namespaceEntry := mutableState.GetNamespaceEntry()
+	namespaceName := namespaceEntry.Name()
+	nextEventID := mutableState.GetNextEventID()
+	closeFailoverVersion, err := mutableState.GetLastWriteVersion()
+	if err != nil {
+		return err
+	}
+
+	executionInfo := mutableState.GetExecutionInfo()
+	workflowTypeName := executionInfo.GetWorkflowTypeName()
+
+	executionState := mutableState.GetExecutionState()
+	memo := getWorkflowMemo(copyMemo(executionInfo.Memo))
+	_, err = e.archiver.Archive(ctx, &archival.Request{
+		ShardID:              e.shardContext.GetShardID(),
+		NamespaceID:          task.NamespaceID,
+		Namespace:            namespaceName.String(),
+		WorkflowID:           task.WorkflowID,
+		RunID:                task.RunID,
+		BranchToken:          branchToken,
+		NextEventID:          nextEventID,
+		CloseFailoverVersion: closeFailoverVersion,
+		HistoryURI:           namespaceEntry.HistoryArchivalState().URI,
+		WorkflowTypeName:     workflowTypeName,
+		StartTime:            executionInfo.GetStartTime(),
+		ExecutionTime:        executionInfo.GetExecutionTime(),
+		CloseTime:            executionInfo.GetCloseTime(),
+		Status:               executionState.Status,
+		HistoryLength:        nextEventID - 1,
+		Memo:                 memo,
+		SearchAttributes:     getSearchAttributes(copySearchAttributes(executionInfo.SearchAttributes)),
+		VisibilityURI:        namespaceEntry.VisibilityArchivalState().URI,
+		Targets:              []archival.Target{archival.TargetHistory, archival.TargetVisibility},
```

Comment:
I think ns can choose to enable only one of history or visibility archival.

Reply:
Good call, fixed!